File: amazon/cousins-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        
        self.x_height = -1
        self.y_height = -2
        
        self.x_parent = None
        self.y_parent = None
        
        def dfs(root, parent, height, x, y):
            
            if root is None: return
            
            if root.val == x:
                self.x_parent = parent
                self.x_height = height
            elif root.val == y:
                self.y_parent = parent
                self.y_height = height
            else:
                dfs(root.left, root, height + 1, x, y)
                dfs(root.right, root, height + 1, x, y)
                
        
        dfs(root, None, 0, x, y)
        
        return self.x_height == self.y_height and self.x_parent != self.y_parent
            

File: amazon/cousins_in_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    TreeNode* x_parent = nullptr;
    TreeNode* y_parent = nullptr;
    
    int x_height = -1;
    int y_height = -2;
    
    void dfs(TreeNode* root, TreeNode* parent, int height, int x, int y) {
        
        if (root == nullptr) return;
        
        if (x == root->val) {
            x_parent = parent;
            x_height = height;
        }
        else if (y == root->val) {
            y_parent = parent;
            y_height = height;
        }
        else {
            dfs(root->left, root, height + 1, x, y);
            dfs(root->right, root, height + 1, x, y);
        }
    }
    
    bool isCousins(TreeNode* root, int x, int y) {
        
        dfs(root, nullptr, 0, x, y);
        
        return (x_height == y_height) && (x_parent != y_parent);
    }
};
